### PR TITLE
Add ServiceLocator and InputManager

### PR DESCRIPTION
**Pull request by Matthew-DeLong**

### CODE REVIEW

File: Assets/Scripts/InputManager/Implementation/RewiredPlayerInputManager.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
+﻿namespace HarmonyQuest.Input
+{ 
+    using System.Collections;
```
Comment (mitjmcc) [inline Assets/Scripts/InputManager/Implementation/RewiredPlayerInputManager.cs]: We need these Collections imports?

File: Assets/Scripts/InputManager/IPlayerInputManager.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+﻿namespace HarmonyQuest.Input
```
Comment (nshooter3) [inline Assets/Scripts/InputManager/IPlayerInputManager.cs]: What's our naming convention for namespaces? Mitch and I just used game name + folder path under the scripts folder. Whatever we decide on, it should be consistent.

File: Assets/Scripts/ServiceLocator.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+        //List of managers the ServiceLocator can provide
+        IPlayerInputManager InputManager;
+        //End List of managers
+       
+        void Awake()
+        {
+            if (instance == null)
+            {
+                instance = this;
```
Comment (nshooter3) [inline Assets/Scripts/ServiceLocator.cs]: See how Mitch and I handle singleton instantiation in our reviews. In order to ensure that script order of operations on awake does not matter, we check for singleton setup both in Awake and in our singleton getters (We can still manually define script orders in unity, but this is a nice failsafe). CameraController.cs has an example of this.

File: Assets/Scripts/ServiceLocator.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+        public IPlayerInputManager GetInputManager()
+        {
+            if(InputManager != null)
+            {
+                return InputManager;
+            }
+            else
+            {
+                Debug.Log("No Input Manager detected to handle request. Starting one up");
```
Comment (nshooter3) [inline Assets/Scripts/ServiceLocator.cs]: This should probably be a Debug.LogWarning.

File: Assets/Scripts/InputManager/IPlayerInputManager.cs (hunk 2 above)
Comment (Matthew-DeLong) [reply Assets/Scripts/InputManager/IPlayerInputManager.cs]: Yeah u rite

File: Assets/Scripts/ServiceLocator.cs (hunk 3 above)
Comment (Matthew-DeLong) [reply Assets/Scripts/ServiceLocator.cs]: Done! This is also nice in that it makes instance readonly.

File: Assets/Scripts/ServiceLocator.cs (hunk 4 above)
Comment (Matthew-DeLong) [reply Assets/Scripts/ServiceLocator.cs]: Done

**Review by mitjmcc — CHANGES_REQUESTED** [review]

**Review by nshooter3 — APPROVED** [review]

File: Assets/Scripts/InputManager/Implementation/RewiredPlayerInputManager.cs (hunk 1 above)
Comment (Matthew-DeLong) [reply Assets/Scripts/InputManager/Implementation/RewiredPlayerInputManager.cs]: Yes. I was saving them for later, but I guess I can remove them for your sake.

**Review by mitjmcc — APPROVED** [review]

**Pull request merged**